### PR TITLE
Authenticate hex.search http requests

### LINKED ISSUE #412: Authenticate hex.search HTTP requests
This will include private packages in the search results. Only do this if there is a user authenticated on the machine.

/cc @Gazler

### PR DESCRIPTION
**Pull request by milmazz**
Closes #412

### CODE REVIEW

File: lib/mix/tasks/hex.search.ex [hunk 1]
```diff
@@ -33,6 +32,19 @@ defmodule Mix.Tasks.Hex.Search do
     end
   end
 
+  defp search_package(package, organization) do
+    result =
+      case Keyword.fetch(Hex.Config.read(), :username) do
```
Comment (milmazz) [inline lib/mix/tasks/hex.search.ex]: @ericmj @wojtekmach Is this the best way to know if there is at least one user authenticated on the machine?

Comment (ericmj) [reply lib/mix/tasks/hex.search.ex]: Probably best to use this check: https://github.com/hexpm/hex/blob/master/lib/mix/tasks/hex.ex#L95-L101.

File: test/mix/tasks/hex.search_test.exs [hunk 2]
```diff
@@ -14,4 +14,21 @@ defmodule Mix.Tasks.Hex.SearchTest do
...
+      send self(), {:mix_shell_input, :yes?, false}
+      send self(), {:mix_shell_input, :prompt, "hunter42"}
+      Mix.Tasks.Hex.Search.run(["doc"])
+
+      assert_received {:mix_shell, :info, ["ex_doc\e[0m" <> ex_doc]}
+      assert_received {:mix_shell, :info, ["only_doc\e[0m" <> only_doc]}
+      assert ex_doc =~ ~r"\w*0\.1\.0.*http://localhost:4043/packages/ex_doc"
+      assert only_doc =~ ~r"\w*0\.1\.0.*http://localhost:4043/packages/only_doc"
+    end
+  end
```
Comment (ericmj) [inline test/mix/tasks/hex.search_test.exs]: Can you add a test for invalid auth credentials also? I think this test would pass even without an authenticated user.

File: lib/mix/tasks/hex.search.ex (hunk 1 above)
Comment (milmazz) [reply lib/mix/tasks/hex.search.ex]: @ericmj Thanks! I already applied your suggestion. Let me know what do you think.

File: test/mix/tasks/hex.search_test.exs (hunk 2 above)
Comment (milmazz) [reply test/mix/tasks/hex.search_test.exs]: @ericmj You're right! The test passes even with invalid authentication credentials. Do you have any suggestion? should we delete both tests?

File: lib/mix/tasks/hex.search.ex (hunk 1 above)
Comment (ericmj) [reply lib/mix/tasks/hex.search.ex]: Sorry, I didn't mean use the function directly and rescue the error. Do the same check that auth_info() does internally or extract to a separate function `authenticated_user?`.

File: test/mix/tasks/hex.search_test.exs (hunk 2 above)
Comment (ericmj) [reply test/mix/tasks/hex.search_test.exs]: Hum. I don't think we can test this properly without a private package and we cant test private packages without being able to create organizations which would require changes on hexpm. Just leave a comment for now that we cant properly test this and we will get to it later.

**Comment by wojtekmach** [comment]
Thank you!

**Pull request merged**